File: variants/java-makefile/net.pijcke.nim.rules/src/main/java/net/pijcke/nim/rules/Nim.java
package net.pijcke.nim.rules;

public class Nim {

  public static final int MAX_STICKS_PER_TURN = 3;

  private Player currentPlayer;
  private Player otherPlayer;

  private int sticks;

  public Nim(final Player player1, final Player player2, final int sticks) {
    this.currentPlayer = player1;
    this.otherPlayer = player2;

    this.sticks = sticks;
  }

  public void start() {
    while (true) {
      final int sticksToRemove = currentPlayer.play(sticks);

      if (sticksToRemove < 1 || sticksToRemove > MAX_STICKS_PER_TURN || sticksToRemove > sticks) {
        broadcast(String.format("Player %s has been eliminated for not following the rules.", currentPlayer));
        break;
      }

      if (sticksToRemove == sticks) {
        broadcast(String.format("Player %s has been eliminated for removing the last stick.", currentPlayer));
        break;
      }

      sticks -= sticksToRemove;
      if (sticks > 1) {
        broadcast(String.format("%s removed %d sticks. There are %d remaining sticks.", currentPlayer, sticksToRemove, sticks));
      }
      else {
        broadcast(String.format("%s removed %d sticks. There is one stick remaining.", currentPlayer, sticksToRemove, sticks));
      }

      swapPlayers();
    }

    broadcast(String.format("The winner is %s!", otherPlayer));
  }

  private void broadcast(String message) {
    currentPlayer.message(message);
    otherPlayer.message(message);
  }

  private void swapPlayers() {
    Player temp = currentPlayer;
    currentPlayer = otherPlayer;
    otherPlayer = temp;
  }

}

File: rules/src/main/kotlin/net/pijcke/nim/rules/Nim.kt
package net.pijcke.nim.rules

class Nim(private val players : List<Player>, private var sticks : Int) {

    private val playing : MutableList<Player> = players.toMutableList()

    companion object {
        const val MAX_STICKS_PER_TURN = 3;
    }

    fun start() {
        while (sticks > 0 && players.count() > 1) {
            val currentPlayer = playing.removeFirst()

            val sticksToRemove = players.first().play(sticks)

            if (sticksToRemove < 1 || sticksToRemove > MAX_STICKS_PER_TURN || sticksToRemove > sticks) {
                broadcast("Player $currentPlayer has been eliminated for not following the rules.")
                continue
            }

            if (sticksToRemove == sticks) {
                broadcast("Player $currentPlayer has been eliminated for removing the last stick.")
                continue
            }

            sticks -= sticksToRemove
            broadcast("$currentPlayer removed $sticksToRemove sticks. There are $sticks remaining sticks.")

            playing.add(currentPlayer)
        }

        if (playing.count() == 1) {
            broadcast("The winner is ${players.first()}!")
        }
        else {
            broadcast("The winners are $players!")
        }
    }

    private fun broadcast(message : String) {
        println(message)

        players.forEach {
            it.message(message)
        }
    }

}